# Customer Churn - Data Cleaning and Model Training

**Usually a client data comes with many polution and bad structurated. In this notebook the goal will be to clean the client data and also train a model on that data. The model will try to predict if the client will or not leave the company.** 

In [1]:
import pandas as pd
df = pd.read_csv("customer_churn.csv")

In [2]:
df.sample(50)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
4911  8565-WUXZU    Male              1     Yes         No      72   
2532  9825-YCXWZ  Female              1      No         No      41   
4154  4229-CZMLL    Male              0      No         No       6   
887   1258-YMZNM  Female              1      No         No      27   
3023  0238-WHBIQ    Male              0     Yes        Yes      72   
2054  9734-UYXQI  Female              0      No         No       2   
2318  6463-HHXJR  Female              0     Yes        Yes      71   
930   5146-CBVOE  Female              0      No         No      16   
6134  3258-SANFR    Male              1      No         No       6   
434   6048-NJXHX    Male              0     Yes         No      41   
222   6469-MRVET    Male              0     Yes        Yes       1   
3082  3401-URHDA    Male              0      No         No      51   
611   4765-OXPPD  Female              0     Yes        Yes       9   
4999  1699-TLDLZ  Female              0     Yes        Yes      16   
3960  7404-JLKQG  Female              0      No         No       3   
5000  5600-PDUJF    Male              0      No         No       6   
3191  7636-OWBPG    Male              1      No         No      12   
1720  6407-UTSLV  Female              1      No         No       2   
5196  0786-IVLAW  Female              1      No         No      66   
4684  6305-YLBMM    Male              0      No         No      69   
3575  9501-UKKNL    Male              0      No         No      59   
4873  2320-YKQBO  Female              0      No         No       7   
1496  3195-TQDZX    Male              0      No         No       3   
5078  9402-ROUMJ  Female              0     Yes        Yes      67   
1394  2272-JKMSI  Female              0     Yes        Yes      62   
3401  4299-OPXEJ  Female              0      No         No      56   
3342  0384-LPITE    Male              0      No         No      40   
4289  2228-BZDEE  Female              0      No         No      54   
4385  1930-BZLHI    Male              0      No         No      21   
1840  6990-YNRIO    Male              0     Yes        Yes      65   
1702  4423-YLHDV  Female              0     Yes         No      21   
3266  0464-WJTKO  Female              0     Yes        Yes      72   
2358  6974-DAFLI  Female              0     Yes         No      55   
2342  1552-AAGRX  Female              0      No         No      30   
6099  7711-GQBZC  Female              0     Yes        Yes      71   
962   9489-UTFKA    Male              1     Yes         No       8   
6784  9546-KDTRB  Female              0      No         No      19   
1370  7596-IIWYC  Female              0      No         No      27   
1898  6689-TCZHQ  Female              1      No         No       5   
2383  3787-TRIAL    Male              0     Yes        Yes      21   
4952  7379-FNIUJ    Male              0      No         No       2   
2313  9154-QDGTH    Male              0     Yes        Yes      43   
1587  5939-SXWHM    Male              0     Yes        Yes      48   
1570  3070-FNFZQ  Female              0     Yes        Yes      23   
3298  1336-EZFZY  Female              0      No         No       4   
4145  8474-UMLNT  Female              0      No         No      13   
1710  6435-SRWBJ  Female              0      No         No       5   
6620  6457-USBER    Male              0      No         No       1   
1713  1702-CCFNJ    Male              0     Yes         No      52   
5230  5887-IKKYO    Male              0     Yes        Yes      58   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
4911          Yes                No     Fiber optic                  Yes  ...   
2532          Yes                No     Fiber optic                   No  ...   
4154          Yes               Yes     Fiber optic                   No  ...   
887           Yes               Yes     Fiber optic                  Yes  ...   
3023          Yes             

**So as we can see this data is not clean. We have some useless data like 'customerID', we have a lot of strings like "Yes" or "No", and for some columns we even have multiple data possibilities, like 'Contract'. Lets start to clean this data the best that we can:**

In [3]:
def print_unique_col_values(df):
       for column in df:
            if df[column].dtypes=='object':
                print(f'{column}: {df[column].unique()}') 

In [4]:
print_unique_col_values(df)

customerID: ['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
gender: ['Female' 'Male']
Partner: ['Yes' 'No']
Dependents: ['No' 'Yes']
PhoneService: ['No' 'Yes']
MultipleLines: ['No phone service' 'No' 'Yes']
InternetService: ['DSL' 'Fiber optic' 'No']
OnlineSecurity: ['No' 'Yes' 'No internet service']
OnlineBackup: ['Yes' 'No' 'No internet service']
DeviceProtection: ['No' 'Yes' 'No internet service']
TechSupport: ['No' 'Yes' 'No internet service']
StreamingTV: ['No' 'Yes' 'No internet service']
StreamingMovies: ['No' 'Yes' 'No internet service']
Contract: ['Month-to-month' 'One year' 'Two year']
PaperlessBilling: ['Yes' 'No']
PaymentMethod: ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
TotalCharges: ['29.85' '1889.5' '108.15' ... '346.45' '306.6' '6844.5']
Churn: ['No' 'Yes']


**With this function we can spot better the strcuture of our table. We can see for exaple which columns are gonna to be just parsed to 0's and 1's and the others that we will need to perform "One Hot Encoding" for example.**

In [5]:
pd.to_numeric(df.TotalCharges)

ValueError: Unable to parse string " " at position 488

**Looks like we have empty rows in TotalCharges column, lets see how many:**

In [6]:
df.shape

(7043, 21)

In [7]:
dfclean = df[df.TotalCharges != ' '].copy()

In [8]:
dfclean.shape

(7032, 21)

**It seems that the overall difference is low(11 in 7043), so we are just gonna remove this rows where the TotalCharge column is blank**

In [9]:
dfclean.drop('customerID', axis = 'columns',inplace=True)

In [10]:
dfclean.TotalCharges = pd.to_numeric(dfclean.TotalCharges)

In [11]:
dfclean.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object

In [12]:
print_unique_col_values(dfclean)

gender: ['Female' 'Male']
Partner: ['Yes' 'No']
Dependents: ['No' 'Yes']
PhoneService: ['No' 'Yes']
MultipleLines: ['No phone service' 'No' 'Yes']
InternetService: ['DSL' 'Fiber optic' 'No']
OnlineSecurity: ['No' 'Yes' 'No internet service']
OnlineBackup: ['Yes' 'No' 'No internet service']
DeviceProtection: ['No' 'Yes' 'No internet service']
TechSupport: ['No' 'Yes' 'No internet service']
StreamingTV: ['No' 'Yes' 'No internet service']
StreamingMovies: ['No' 'Yes' 'No internet service']
Contract: ['Month-to-month' 'One year' 'Two year']
PaperlessBilling: ['Yes' 'No']
PaymentMethod: ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn: ['No' 'Yes']


In [13]:
dfclean.replace('No internet service','No', inplace=True)
dfclean.replace('No phone service','No', inplace=True)

**Now that we have a big quantity of binary possibilities columns lets change those strings to proper numbers. The replaces above could be threated with one hot enconding but for simplicity of the exercise we gonna threat they like 'No' to.**

In [14]:
binary_columns = ['Partner','Dependents','PhoneService','MultipleLines','OnlineSecurity','OnlineBackup',
                  'DeviceProtection','TechSupport','StreamingTV','StreamingMovies','PaperlessBilling']
for col in binary_columns:
    dfclean[col].replace({'Yes':1, 'No':0}, inplace=True)

In [15]:
dfclean.gender.replace({'Female':1,'Male':0},inplace=True)

In [16]:
print_unique_col_values(dfclean)

InternetService: ['DSL' 'Fiber optic' 'No']
Contract: ['Month-to-month' 'One year' 'Two year']
PaymentMethod: ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn: ['No' 'Yes']


In [17]:
dfclean.dtypes

gender                int64
SeniorCitizen         int64
Partner               int64
Dependents            int64
tenure                int64
PhoneService          int64
MultipleLines         int64
InternetService      object
OnlineSecurity        int64
OnlineBackup          int64
DeviceProtection      int64
TechSupport           int64
StreamingTV           int64
StreamingMovies       int64
Contract             object
PaperlessBilling      int64
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object

**Now we already have most of our data clean, only the multi-class columns are remaining. Let's use one hot enconding for those columns:**

In [18]:
dfclean = pd.get_dummies(data = dfclean,columns=['InternetService','Contract','PaymentMethod'])

In [19]:
dfclean.sample(10)

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
6478       1              0        1           0      60             1   
861        0              0        1           1      64             1   
1840       0              0        1           1      65             1   
1874       1              0        1           0      43             1   
1050       0              1        0           0      12             0   
5358       0              0        0           0      18             1   
5441       1              0        1           0      70             1   
1025       1              1        1           0      60             1   
4446       0              1        0           0       3             1   
5250       0              0        0           0      12             1   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
6478              1               0             1                 1  ...   
861               1               1             0                 1  ...   
1840              1               1             0                 1  ...   
1874              1               1             1                 0  ...   
1050              0               0             0                 1  ...   
5358              0               0             0                 0  ...   
5441              1               1             0                 1  ...   
1025              0               1             1                 1  ...   
4446              0               0             0                 0  ...   
5250              1               0             0                 0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
6478                    0                            1                   0   
861                     0                            1                   0   
1840                    0                            1                   0   
1874                    1                            0                   0   
1050                    1                            0                   0   
5358                    0                            0                   1   
5441                    1                            0                   0   
1025                    0                            1                   0   
4446                    0                            1                   0   
5250                    0                            1                   0   

      Contract_Month-to-month  Contract_One year  Contract_Two year  \
6478                        0                  0                  1   
861                         0                  0                  1   
1840                        0                  0                  1   
1874                        0                  0                  1   
1050                        1                  0                  0   
5358                        1                  0                  0   
5441                        0                  0                  1   
1025                        0                  0                  1   
4446                        1                  0                  0   
5250                        1                  0                  0   

     PaymentMethod_Bank transfer (automatic)  \
6478                                       0   
861                                        1   
1840                                       1   
1874                                       0   
1050                                       0   
5358                                       0   
5441                                       0   
1025                                       1   
4446                                       0   
5250                                       0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
6478                                      1                               0   
861                       

**Now our features are in numeric format only. But they are not scaled yet, let's scale then in a way to better train our model further on:**

In [20]:
cols_to_scale = ['tenure','MonthlyCharges','TotalCharges']

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

dfclean[cols_to_scale] = scaler.fit_transform(dfclean[cols_to_scale])

In [21]:
dfclean.sample(10)

gender  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
942        0              0        0           0  0.323944             1   
120        1              0        1           1  0.774648             1   
6875       1              0        1           0  0.887324             1   
1856       1              1        0           0  0.521127             1   
86         1              0        1           0  0.478873             1   
1111       1              0        0           0  0.042254             1   
5521       0              0        0           0  0.816901             1   
6930       1              0        1           0  0.028169             1   
5222       1              0        0           0  0.098592             1   
6554       1              0        0           0  0.000000             1   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
942               0               0             0                 0  ...   
120               1               0             0                 1  ...   
6875              0               0             0                 0  ...   
1856              0               1             1                 0  ...   
86                0               1             0                 0  ...   
1111              0               1             0                 1  ...   
5521              0               0             0                 1  ...   
6930              1               0             0                 0  ...   
5222              1               1             1                 1  ...   
6554              0               0             0                 0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
942                     0                            1                   0   
120                     0                            1                   0   
6875                    0                            0                   1   
1856                    1                            0                   0   
86                      1                            0                   0   
1111                    0                            1                   0   
5521                    1                            0                   0   
6930                    0                            1                   0   
5222                    1                            0                   0   
6554                    0                            1                   0   

      Contract_Month-to-month  Contract_One year  Contract_Two year  \
942                         1                  0                  0   
120                         0                  1                  0   
6875                        0                  0                  1   
1856                        1                  0                  0   
86                          0                  1                  0   
1111                        1                  0                  0   
5521                        1                  0                  0   
6930                        1                  0                  0   
5222                        1                  0                  0   
6554                        1                  0                  0   

     PaymentMethod_Bank transfer (automatic)  \
942                                        0   
120                                        0   
6875                                       0   
1856                                       0   
86                                         1   
1111                                       1   
5521                                       0   
6930                                       0   
5222                                       1   
6554                                       0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
942                                       1                               0   
120 

**Here are the final data to us to use in our model:**

In [22]:
X = dfclean.drop('Churn',axis='columns')                #features
Y = dfclean['Churn']                                    #result
Y.replace({'Yes':1,'No':0},inplace=True)
Y.sample(10)

6302    1
6921    0
3435    0
5133    1
1413    0
94      0
661     0
4769    0
4089    1
789     0
Name: Churn, dtype: int64

# Model Training

**Now that we have our data cleaned let's train a model in order to get some prediction for possible Churn in other clients:**

In [39]:
import tensorflow as tf
from tensorflow import keras

In [40]:
X.shape

(7032, 26)

In [41]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,train_size=0.2)

In [42]:
model = keras.Sequential([
    keras.layers.Dense(50, input_shape=(26,), activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(20,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']    
)

In [43]:
model.fit(X_train,Y_train,epochs=50)

Epoch 1/50
44/44 [==============================] - 1s 884us/step - loss: 0.5650 - accuracy: 0.7283
Epoch 2/50
44/44 [==============================] - 0s 835us/step - loss: 0.4998 - accuracy: 0.7504
Epoch 3/50
44/44 [==============================] - 0s 858us/step - loss: 0.4791 - accuracy: 0.7632
Epoch 4/50
44/44 [==============================] - 0s 858us/step - loss: 0.4540 - accuracy: 0.7817
Epoch 5/50
44/44 [==============================] - 0s 835us/step - loss: 0.4636 - accuracy: 0.7696
Epoch 6/50
44/44 [==============================] - 0s 877us/step - loss: 0.4577 - accuracy: 0.7824
Epoch 7/50
44/44 [==============================] - 0s 858us/step - loss: 0.4450 - accuracy: 0.7831
Epoch 8/50
44/44 [==============================] - 0s 848us/step - loss: 0.4510 - accuracy: 0.7852
Epoch 9/50
44/44 [==============================] - 0s 858us/step - loss: 0.4357 - accuracy: 0.7888
Epoch 10/50
44/44 [==============================] - 0s 877us/step - loss: 0.4427 - accuracy: 0.7781

In [46]:
yp = model.predict(X_test)
Y_pred = []
for element in yp:
    if element > 0.5:
        Y_pred.append(1)
    else:
        Y_pred.append(0)
        
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(Y_test,Y_pred))

176/176 [==============================] - 0s 608us/step
              precision    recall  f1-score   support

           0       0.83      0.90      0.87      4137
           1       0.64      0.51      0.57      1489

    accuracy                           0.79      5626
   macro avg       0.74      0.70      0.72      5626
weighted avg       0.78      0.79      0.79      5626



**With the data cleanned we could construct this model with a f1 score of 79%**